src/Orchard.Cms.Web/Modules/Orchard.OpenId/Controllers/AccessController.cs

```diff
@@ -48,12 +54,25 @@ public class AccessController : Controller
             // which is required for this stateless OAuth2/OIDC token endpoint to work correctly.
             // To prevent effective CSRF/session fixation attacks, this action MUST NOT
             // return an authentication cookie or try to establish an ASP.NET Core session.
+            var openIdSettings = await _openIdService.GetOpenIdSettingsAsync();
+            var application = await _applicationManager.FindByClientIdAsync(request.ClientId);
+            if (application == null)
+            {
+                return View("Error", new ErrorViewModel
+                {
+                    Error = OpenIdConnectConstants.Errors.InvalidClient,
+                    ErrorDescription = T["Details concerning the calling client application cannot be found in the database"]
+                });
+            }
+
+            var isPasswordGrantType = request.IsPasswordGrantType() && openIdSettings.AllowPasswordFlow && application != null && application.AllowPasswordFlow;
```

Comment:
Sounds like some utterly complicated logic, IMHO :scream: 

And the bad thing is that it doesn't reject the request with an appropriate error message when the flow can't be used by the application.

Note: you don't need to check whether `openIdSettings.AllowXYZFlow` is true or not at this point, as OpenIddict will immediately reject the request if it uses a disabled flow (this action won't be invoked). Consider using debug asserts if you want to keep a sanity check somewhere:

```
Debug.Assert(request.IsPasswordGrantType() && !openIdSettings.AllowPasswordFlow,
    "Password grant requests should be rejected by OpenIddict when the corresponding flow is not enabled.");

Debug.Assert(request.IsClientCredentialsGrantType() && !openIdSettings.AllowClientCredentialsFlow,
    "Password grant requests should be rejected by OpenIddict when the corresponding flow is not enabled.");